### ISSUE TITLE
Ownership of nested resources is inconsistent, breaks backwards compatibility

### ISSUE THREAD

**Issue by irisjae** [onflow/cadence#1320]
### Problem

Flow seems to handle the `.owner` of nested resources inconsistently, and appears to have made a breaking change in `.owner` behaviour in the most recent (Dec 8, 2021) spork.

[One would expect](https://docs.onflow.org/cadence/language/composite-types/#resource-owner) that the `.owner` of a nested resource should always be pointing to the latest account which it had been stored in, and always be the same as its nesting resource. However, Flow inconsistently exhibits unintuitive behaviour.

Before the Dec 8 spork (emulator version 0.28.1), the `.owner` of a nested resource is consistently the same as its nesting resource. However, if a resource is stored in an array, which is in an account's storage, the resource's `.owner` will not be recorded and remain `nil`.

After the Dec 8 spork (emulator version 0.30.1),  the `.owner` of a nested resource is not recorded and remains `nil` even when its nesting resource's `.owner` changes.

### Steps to Reproduce

This can be verified with a test contract and transaction, as follows:
```
// TestContract.cdc
pub contract TestContract {
        pub resource TestResource {}
        pub resource TestNestingResource {
                pub let testResources : @[TestResource]

                init () {
                        self .testResources <- [ <- create TestResource () ] }
                destroy () {
                        destroy self .testResources } }
        pub fun makeTestResource () : @TestResource {
                return <- create TestResource () }
        pub fun makeTestNestingResource () : @TestNestingResource {
                return <- create TestNestingResource () }

        pub event Log (info : String)
        pub fun log (_ info : String) : Void {
                emit Log (info: info) } }
```

```
// test-transaction.cdc
import TestContract from 0xTESTCONTRACT

transaction () {

        prepare (account : AuthAccount) {

                var nestingResource <- TestContract .makeTestNestingResource ()
                var nestingResourceRef = & nestingResource as &TestContract.TestNestingResource
                var nestedElementResourceRef = & (nestingResource .testResources [0]) as &TestContract.TestResource

                // At this point the nesting and nested resources are not in storage
                TestContract .log ("nesting resource owner: " .concat (nestingResourceRef .owner ?.address ?.toString () ?? "nil"))
                TestContract .log ("nested resource owner: " .concat (nestedElementResourceRef .owner ?.address ?.toString () ?? "nil"))

                account .save (<- nestingResource, to: /storage/test)

                // At this point the nesting and nested resources are both in storage
                TestContract .log ("nesting resource owner: " .concat (nestingResourceRef .owner ?.address ?.toString () ?? "nil"))
                TestContract .log ("nested resource owner: " .concat (nestedElementResourceRef .owner ?.address ?.toString () ?? "nil"))

                var nestingResource2 <- account .load <@TestContract.TestNestingResource> (from: /storage/test) !

                // At this point the nesting and nested resources are both taken out of storage
                TestContract .log ("nesting resource owner: " .concat (nestingResourceRef .owner ?.address ?.toString () ?? "nil"))
                TestContract .log ("nested resource owner: " .concat (nestedElementResourceRef .owner ?.address ?.toString () ?? "nil"))

                var removedResource <- nestingResource2 .testResources .remove (at: 0)

                // At this point the nesting and nested resources are both still out of storage, and the nested resource is removed from the nesting resource
                TestContract .log ("nesting resource owner: " .concat (nestingResourceRef .owner ?.address ?.toString () ?? "nil"))
                TestContract .log ("nested resource owner: " .concat (nestedElementResourceRef .owner ?.address ?.toString () ?? "nil"))

                destroy removedResource
                destroy nestingResource2 !

                var resourceArray <- [ <- TestContract .makeTestResource () ]
                var elementResourceRef = & (resourceArray [0]) as &TestContract.TestResource

                // At this point the resource array and the nested resource are not in storage
                TestContract .log ("element resource owner: " .concat (elementResourceRef .owner ?.address ?.toString () ?? "nil"))

                account .save (<- resourceArray, to: /storage/test)

                // At this point the resource array and the nested resource are both in storage
                TestContract .log ("element resource owner: " .concat (elementResourceRef .owner ?.address ?.toString () ?? "nil"))

                var resourceArray2 <- account .load <@[TestContract.TestResource]> (from: /storage/test) !

                // At this point the resource array and the nested resource are both taken out of storage
                TestContract .log ("element resource owner: " .concat (elementResourceRef .owner ?.address ?.toString () ?? "nil"))

                var removedResource2 <- resourceArray2 .remove (at: 0)

                // At this point the resource array and the nested resource are both still out of storage, and the nested resource is removed from the resource array
                TestContract .log ("element resource owner: " .concat (elementResourceRef .owner ?.address ?.toString () ?? "nil"))

                destroy removedResource2
                destroy resourceArray2 } }
```

Before the Dec 8 spork (emulator version 0.28.1), this script gives the result by inspecting events:
```
nesting resource owner: nil
nested resource owner: nil
nesting resource owner: 0xf8d6e0586b0a20c7
nested resource owner: 0xf8d6e0586b0a20c7
nesting resource owner: 0xf8d6e0586b0a20c7
nested resource owner: 0xf8d6e0586b0a20c7
nesting resource owner: 0xf8d6e0586b0a20c7
nested resource owner: 0xf8d6e0586b0a20c7
element resource owner: nil
element resource owner: 0xf8d6e0586b0a20c7
element resource owner: nil
element resource owner: nil
```

After the Dec 8 spork (emulator version 0.30.1), this script gives the result by inspecting events:
```
nesting resource owner: nil
nested resource owner: nil
nesting resource owner: 0xf8d6e0586b0a20c7
nested resource owner: nil
nesting resource owner: 0xf8d6e0586b0a20c7
nested resource owner: nil
nesting resource owner: 0xf8d6e0586b0a20c7
nested resource owner: nil
element resource owner: nil
element resource owner: nil
element resource owner: nil
element resource owner: nil
```

**Comment by irisjae** [comment]
This affects use cases such as checking the owner of NFTs or other resources stored in collections.

**Comment by irisjae** [comment]
It seems this was [introduced](https://forum.onflow.org/t/breaking-change-bugfix-for-cadence-resource-owner-field/2576) as a bugfix. Nonetheless, the behaviour for nested resources does not seem correct.

**Comment by turbolent** [comment]
Thank you for reporting this bug! 

Indeed, the behavior of the `owner` field was fixed in [v0.20](https://github.com/onflow/cadence/releases/tag/v0.20.0), however, it looks like the behavior of references to dictionary elements was not fixed properly.

Thank you for providing a reproduction. I am able to reproduce this locally using a fixed subset of it.
